File: src/helpers/loadfile.zig
// Copyright (c) 2021 VisualDevelopment. All rights reserved.
const uefi = @import("std").os.uefi;
const assert = @import("assert.zig").assert;
const con_out_writer = @import("conoutwriter.zig").con_out_writer;

pub var esp: *uefi.protocols.FileProtocol = undefined;

pub fn openESP() void {
    const boot_services = uefi.system_table.boot_services.?;
    var image: ?*uefi.protocols.LoadedImageProtocol = undefined;
    var fs: ?*uefi.protocols.SimpleFileSystemProtocol = undefined;
    assert(boot_services.handleProtocol(uefi.handle, &uefi.protocols.LoadedImageProtocol.guid, @ptrCast(*?*c_void, &image)), "Unable to open image info.", @src());
    assert(boot_services.handleProtocol(image.?.device_handle.?, &uefi.protocols.SimpleFileSystemProtocol.guid, @ptrCast(*?*c_void, &fs)), "Unable to open filesystem", @src());
    assert(fs.?.openVolume(&esp), "Unable to open volume.", @src());
}

pub const LoadFileError = error{
    FileOpenFailure,
    FileInfoFailure,
    AllocationFailure,
    ReadFailure,
};

pub fn loadFile(path: [:0]const u16, open_mode: u64, attributes: u64) LoadFileError![]align(8) u8 {
    // Open file
    var file: *uefi.protocols.FileProtocol = undefined;
    switch (esp.open(&file, path, open_mode, attributes)) {
        .Success => {},
        else => |sts| {
            con_out_writer.print("Failed to open file handle: {}", .{sts}) catch unreachable;
            return LoadFileError.FileOpenFailure;
        },
    }

    // Get file size
    var info_buffer_size: usize = 0;
    var info_buffer: *uefi.protocols.FileInfo = undefined;
    _ = file.getInfo(&uefi.protocols.FileProtocol.guid, &info_buffer_size, @ptrCast([*]u8, info_buffer));
    assert(info_buffer_size > 0, "File too small", @src());
    switch (uefi.system_table.boot_services.?.allocatePool(uefi.tables.MemoryType.LoaderData, info_buffer_size, @ptrCast(*[*]align(8) u8, &info_buffer))) {
        .Success => {},
        else => |sts| {
            con_out_writer.print("Failed to allocate memory for file info: {}", .{sts}) catch unreachable;
            return LoadFileError.AllocationFailure;
        },
    }
    switch (file.getInfo(&uefi.protocols.FileProtocol.guid, &info_buffer_size, @ptrCast([*]u8, info_buffer))) {
        .Success => {},
        else => |sts| {
            con_out_writer.print("Failed to get file info: {}", .{sts}) catch unreachable;
            return LoadFileError.FileInfoFailure;
        },
    }

    // Read file data
    var buffer_size: usize = info_buffer.file_size;
    _ = uefi.system_table.boot_services.?.freePool(@ptrCast([*]align(8) u8, &info_buffer));
    var buffer: [*]align(8) u8 = undefined;
    switch (uefi.system_table.boot_services.?.allocatePool(uefi.tables.MemoryType.LoaderData, buffer_size, &buffer)) {
        .Success => {},
        else => |sts| {
            con_out_writer.print("Failed to allocate file buffer: {}", .{sts}) catch unreachable;
            return LoadFileError.AllocationFailure;
        },
    }
    switch (file.read(&buffer_size, buffer)) {
        .Success => {},
        else => |sts| {
            con_out_writer.print("Failed to read file data: {}", .{sts}) catch unreachable;
            return LoadFileError.ReadFailure;
        },
    }
    // Close file
    _ = file.close();

    return buffer[0..buffer_size];
}

File: src/helpers/load_file.rs
use ::alloc::*;
use uefi::prelude::*;
use uefi::proto::media::file::{Directory, File, FileAttribute, FileInfo, FileMode, FileType};

static mut ESP: Option<Directory> = None;

pub fn open_esp(image: Handle) {
    unsafe {
        let fs = uefi_services::system_table().as_mut().boot_services().get_image_file_system(image).unwrap().unwrap().get().as_mut().unwrap();
        ESP = Some(fs.open_volume().unwrap().unwrap());
    }
}

pub fn load_file(path: &str) -> vec::Vec<u8, alloc::Global> {
    let esp = unsafe { ESP.as_mut().unwrap() };
    let mut file = match esp
        .open(path, FileMode::Read, FileAttribute::empty())
        .expect_success(format!("File {} not found", path).as_str())
        .into_type()
        .unwrap()
        .unwrap()
    {
        FileType::Regular(f) => f,
        _ => panic!("How do you expect me to load the {} folder?", path),
    };

    let mut buffer = vec![
        0;
        file.get_boxed_info::<FileInfo>()
            .expect_success(format!("Failed to get {} file info", path).as_str())
            .file_size()
            .try_into()
            .unwrap()
    ];
    file.read(&mut buffer).expect_success(format!("Failed to read {}.", path).as_str());

    buffer
}
